### PR TITLE
FEATURE: BB-1856, BB-1857

### PR DESCRIPTION
**Pull request by strouhaljan**
- BB-1856 target measure in Bullet chart
- BB-1857 comparative measure in Bullet chart

---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [ ] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [ ] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [ ] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [ ] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Successful `extended test - examples`
- [ ] Successful `extended test - storybook`
- [ ] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer:
- gdc-dashboards:

# Related Jira tasks

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5761/ : SUCCESS in 9m 33s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1315/ : SUCCESS in 8s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5766/ : SUCCESS in 12m 06s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1322/ : SUCCESS in 7s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2945/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1323/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2359/ : UNSTABLE in 16m 24s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5768/ : SUCCESS in 9m 58s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1325/ : SUCCESS in 5s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2947/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1327/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2360/ : UNSTABLE in 21m 31s

**Comment by strouhaljan** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1341.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1330/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2362/ : SUCCESS in 22m 03s

File: src/components/visualizations/chart/HighChartsRenderer.tsx [hunk 1]
```diff
@@ -246,15 +246,19 @@ export default class HighChartsRenderer extends React.PureComponent<
...
         }
 
+        const interactive = !isBulletChart(type);
+
         const legendProps = {
             position: legend.position,
             responsive: legend.responsive,
             chartType: type,
             series: items,
+            interactive,
```
Comment (tomassatinsky) [inline src/components/visualizations/chart/HighChartsRenderer.tsx]: Please verify with Petr if this is needed

File: src/components/visualizations/chart/chartOptions/bulletChartOptions.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+const getTargetSeries = (seriesItemConfig: IPointData) => ({
+    ...seriesItemConfig,
+    type: "bullet",
+    pointPadding: 0,
+    zIndex: 2,
+});
+
+const getComparativeSeries = (seriesItemConfig: IPointData) => ({
+    ...seriesItemConfig,
+    pointPadding: -0.2,
```
Comment (tomassatinsky) [inline src/components/visualizations/chart/chartOptions/bulletChartOptions.ts]: have you tried it with a large number of bars doesn't it lead to bar overlaps?

File: stories/core_components/BulletChart.tsx [hunk 3]
```diff
@@ -50,12 +51,45 @@ storiesOf("Core components/BulletChart", module)
...
+                    targetMeasure={MEASURE_2}
+                    viewBy={[ATTRIBUTE_1]}
+                    onError={onErrorHandler}
+                    LoadingComponent={null}
+                    ErrorComponent={null}
+                />
+            </div>,
+        ),
+    )
+    .add("primary, target and comparative measures, one attribute", () =>
```
Comment (tomassatinsky) [inline stories/core_components/BulletChart.tsx]: please add primary and comparative

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5800/ : SUCCESS in 8m 06s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1370/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5820/ : SUCCESS in 9m 58s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1405/ : SUCCESS in 5s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2991/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1409/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2402/ : UNSTABLE in 16m 29s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5826/ : SUCCESS in 11m 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1413/ : SUCCESS in 5s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2995/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1415/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2405/ : UNSTABLE in 13m 32s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5829/ : SUCCESS in 5m 46s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1416/ : SUCCESS in 5s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2997/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1417/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2406/ : SUCCESS in 13m 46s

File: src/components/visualizations/chart/chartOptions/bulletChartOptions.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import { IPointData, ISeriesItem } from "../../../../interfaces/Config";
+import { unwrap } from "../../../../helpers/utils";
+import { IColorStrategy } from "../colorFactory";
+import { SECONDARY_MEASURES, MEASURES, TERTIARY_MEASURES } from "../../../../constants/bucketNames";
+import { filterOutEmptyBuckets } from "../../../../helpers/mdObjBucketHelper";
+
+export type IBucketLocalIdentifier = VisualizationObject.IBucket["localIdentifier"];
+
+const isTargetMeasurePresent = (bucketLocalIdentifiers: IBucketLocalIdentifier[]) =>
+    bucketLocalIdentifiers.indexOf(SECONDARY_MEASURES) !== -1;
```
Comment (yanlipnican) [inline src/components/visualizations/chart/chartOptions/bulletChartOptions.ts]: bucketLocalIdentifiers.includes(SECONDARY_MEASURES)

File: src/components/visualizations/chart/chartOptions/bulletChartOptions.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            ...getValue(pointValue, seriesIndex, bucketsLocalIdentifiers),
+            format: unwrap(measureGroup.items[seriesIndex]).format,
+            marker: {
+                enabled: pointValue !== null,
+            },
+            name: unwrap(measureGroup.items[seriesIndex]).name,
+        };
+
+        return pointData;
+    });
```
Comment (yanlipnican) [inline src/components/visualizations/chart/chartOptions/bulletChartOptions.ts]:
```
    seriesItem.map((pointValue: string) => ({
            ...getValue(pointValue, seriesIndex, bucketsLocalIdentifiers),
            format: unwrap(measureGroup.items[seriesIndex]).format,
            marker: {
                enabled: pointValue !== null,
            },
            name: unwrap(measureGroup.items[seriesIndex]).name,
        }));
```

Comment (yanlipnican) [reply src/components/visualizations/chart/chartOptions/bulletChartOptions.ts]: And add `IPointData` to return type of `getSeriesItemData`

File: src/components/visualizations/chart/chartOptions/bulletChartOptions.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+});
+
+const getTargetSeriesIndex = (bucketsLocalIdentifiers: IBucketLocalIdentifier[]) => {
+    if (!isTargetMeasurePresent(bucketsLocalIdentifiers)) {
+        return null;
+    }
+
+    return (
+        bucketsLocalIdentifiers.filter(
+            bucketLocalIdentifier => [MEASURES, SECONDARY_MEASURES].indexOf(bucketLocalIdentifier) !== -1,
```
Comment (yanlipnican) [inline src/components/visualizations/chart/chartOptions/bulletChartOptions.ts]: includes

File: src/components/visualizations/chart/chartOptions/bulletChartOptions.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+        series[seriesIndex] = getSeries(
+            seriesIndex,
+            seriesItemConfig,
+            notEmptyMeasureBucketsLocalIdentifiers,
+            colorStrategy,
+        );
+    });
+
+    return compact(series);
```
Comment (yanlipnican) [inline src/components/visualizations/chart/chartOptions/bulletChartOptions.ts]: Make it compact

File: src/components/visualizations/chart/chartOptions/test/bulletChartOptions.spec.ts [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+                pointPadding: 0,
+                zIndex: 0,
+            },
+        ];
+
+        const series = getBulletChartSeries(executionResultData, measureGroup, colorStrategy, buckets);
+
+        expect(series).toEqual(expectedSeries);
+    });
+});
```
Comment (yanlipnican) [inline src/components/visualizations/chart/chartOptions/test/bulletChartOptions.spec.ts]: Possibly do the snapshot tests

File: src/components/visualizations/chart/colorFactory.ts [hunk 9]
```diff
@@ -582,6 +591,46 @@ export class ScatterPlotColorStrategy extends PointsChartColorStrategy {
...
+
+    protected createPalette(colorPalette: IColorPalette, colorAssignment: IColorAssignment[]): string[] {
+        if (TypeGuards.isGuidColorItem(colorAssignment[0].color)) {
+            return this.getCustomBulletChartColorPalette(
+                getColorByGuid(colorPalette, colorAssignment[0].color.value as string, 0),
+            );
+        }
+
+        return this.getCustomBulletChartColorPalette(colorAssignment[0].color.value as IColor);
+    }
```
Comment (yanlipnican) [inline src/components/visualizations/chart/colorFactory.ts]: create local variable for `colorAssignment[0].color` it's used 3 times here

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5838/ : SUCCESS in 5m 35s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1425/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5839/ : SUCCESS in 6m 03s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1426/ : SUCCESS in 4s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5840/ : SUCCESS in 4m 54s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1427/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5841/ : SUCCESS in 5m 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1428/ : SUCCESS in 7s

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3005/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1429/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2410/ : SUCCESS in 13m 41s

**Comment by yanlipnican** [comment]
LGTM

**Comment by yanlipnican** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3016/ : SUCCESS in 13m 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5859/ : SUCCESS in 5m 04s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1411/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1953/ : SUCCESS in 21m 51s

**Pull request merged**